files/image_config/config-setup/config-setup

```diff
@@ -0,0 +1,370 @@
+  CONFIG_TYPE=$1
+  DEST_FILE=$2
+
+  if [ "$1" = "ztp" ]; then
+    /usr/lib/ztp/ztp-profile.sh create ${DEST_FILE}
+  elif [ "$1" = "factory" ]; then
+    rv=1
+
+    # Execute config initialization hooks
+    run_hookdir ${FACTORY_DEFAULT_HOOKS} ${CONFIG_SETUP_INITIALIZATION_FLAG}
+
+    # Use preset defined in default_sku
+    if [ ! -e ${DEST_FILE} ]; then
+       sonic-cfggen -H -k ${HW_KEY} --preset ${DEFAULT_PRESET} > ${DEST_FILE}
+       rv=$?
+       if [ $rv -ne 0 ]; then
+          return $rv
+       fi
+    fi
+  fi
+  return 0
+}
+
+# Create SONiC configuration for first time bootup
+#  - If ZTP is enabled, ZTP configuraion is created
+#  - If ZTP is disabled and updategraph is disabled, factory default configuration
+#    is created
+#  - If updategraph is enabled and ZTP is disabled, updategraph initializes
+#    configuration
+do_config_intialization()
+{
+    if  ! updategraph_is_enabled ; then
+        if ! ztp_is_enabled ; then
+           echo "No configuration detected, generating factory default configuration..."
+           generate_config factory ${CONFIG_DB_JSON}
+           load_config ${CONFIG_DB_JSON}
+        fi
+    fi
+
+    if  ztp_is_enabled ; then
+        echo "No configuration detected, initiating zero touch provisioning..."
+        generate_config ztp ${TMP_ZTP_CONFIG_DB_JSON}
+        load_config ${TMP_ZTP_CONFIG_DB_JSON}
+        rm -f ${TMP_ZTP_CONFIG_DB_JSON}
+    fi
+
+    rm -f /tmp/pending_config_initialization
+}
+
+# Perform configuration migration from backup copy.
+#  - This step is performed when a new image is installed and SONiC switch boots into it
+do_config_migration()
+{
+  # Identify list of files to migrate
+  copy_list="minigraph.xml snmp.yml acl.json config_db.json frr"
+
+  # Migrate all configuration files from old to new
+  copy_config_files_and_directories $copy_list
+
+  # Execute custom hooks if present
+  run_hookdir ${CONFIG_POST_MIGRATION_HOOKS} ${CONFIG_SETUP_POST_MIGRATION_FLAG}
+
+  if [ x"${WARM_BOOT}" == x"true" ]; then
+      echo "Warm reboot detected..."
+      disable_updategraph
+      rm -f /tmp/pending_config_migration
+      exit 0
+  elif updategraph_is_enabled ; then
+      if [ -e ${MINGRAPH_FILE} ]; then
+          echo "Use minigraph.xml from old system..."
+          reload_minigraph
+          sonic-cfggen -d --print-data > ${CONFIG_DB_JSON}
+
+          # Disable updategraph
+          disable_updategraph
+      fi
+  else
+      if [ -e ${CONFIG_DB_JSON} ]; then
+         echo "Use config_db.json from old system..."
+         sonic-cfggen -j ${CONFIG_DB_JSON} --write-to-db
+
+         if [[ -x /usr/bin/db_migrator.py ]]; then
+             # Migrate the DB to the latest schema version if needed
+             /usr/bin/db_migrator.py -o migrate
+         fi
+      fi
+  fi
+
+  rm -f /tmp/pending_config_migration
+}
+
+# Take a backup of current SONiC configuration
+do_config_backup()
+{
+  echo "Taking backup of curent configuration"
```

Comment:
can you make tabstop=4?

Reply:
Will do. Waiting for more comments, before I make this cosmetic change.